First draft of section on Tracy

theWatchmen opened the pull request.

theWatchmen commented:
@dendibakh I haven't built the book yet to check the layout. Just wanted to make sure you were happy with the content first :) Let me know if it needs more details.

dendibakh commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
````diff
@@ -0,0 +1,32 @@
+
+We are going to focus on Tracy, as this seems to be one the more popular projects, however the concepts apply to the other profilers as well. Profiling a section of code is really simple: you just need to import Tracy's header and mark the function to be profiled:
+```c++
+void my_function()
````
```suggestion
void render_frame()
```
To make it more realistic :)

dendibakh replied:
Actually, it would be good to show the piece of the real code (if you can) of the project that you profile later in the chapter. So that the code and screenshots match. 5-6 lines of code would be enough. Basically, show how you insert the ZoneScoped macro into a real project.

dendibakh commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -0,0 +1,32 @@
+## Tracing profilers
+Most of the tools explored so far fall under the category of sampling profilers. These are great when you want to identify hot-spots in your code, but in some cases they might not provide enough granularity for analysis. Depending on the profiler sampling frequency and the behavior of your program, most functions could be fast enough that they don't show up in a profiler. In some scenarios you mi ...
+
```
After reading, I still don't understand how exactly Tracy works. It also does sampling, correct? Is it just a higher sampling frequency for the recorded zone?
I also notice Tracy records thread migration and idle time. I wonder how it does that? KUTrace, for example, can do that because it captures every kernel <-> user space transition, but it requires additional Linux kernel patching.

theWatchmen replied:
I will add some details to the explanation. That's correct, Tracy also does system sampling by default, but I am don't exactly how it accomplishes that. I'll ask the author of the library if he can provide more details :)

dendibakh replied:
@wolfpld, how do you trace thread migration and idle time?

wolfpld replied:
Kernel context switch tracing.

dendibakh replied:
Do you have to have the administrator's privilege to do that?

wolfpld replied:
Yes. Non-elevated processes do not have insight into the inner workings of the operating system.

dendibakh commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
````diff
@@ -0,0 +1,32 @@
+    ...
+}
+```
+The `ZoneScoped` macro will create an object on the stack with an associated timestamp. Once the function exits, the object will record a new timestamp and will store this timing data along with a few other details, including the function name. The implementation has very little overhead, although care must be taken when profiling sections of the code like tight loops.
````
Can you please define the overhead? Can we give a percentage range?

theWatchmen replied:
On a synthetic test, they measured the overhead to be 2.25ns. I'll add this to the description.

wolfpld replied:
The test was not artificial.

![obraz](https://user-images.githubusercontent.com/600573/224343014-8bd885c2-35a5-4b2b-85f4-7128705d1dc4.png)

wolfpld replied:
Oh, and specifying the percentage range makes no sense because it is completely up to you how much instrumentation you want to insert. It's a range from 0 to 100%.

dendibakh replied:
Ok, we can still give readers an intuition by providing an example of a typical (default) usage of Tracy. For instance, running app X with features Y and Z enabled and default sampling frequency F, gives N percent of runtime overhead.

dendibakh replied:
@wolfpld, is Tracy a CPU-agnostic tool? In other words, does it leverage the HW PMU features (like PEBS on Intel)? HW-based sampling will significantly lower the runtime overhead (up to 30x).

wolfpld replied:
Tracy takes advantage of the sampling capabilities of the operating system. Under Linux, you also get samples of common microarchitectural counters that are not CPU vendor or CPU architecture dependent (i.e., you get the same statistics on ARM). The implementation of how this is done is outside of Tracy's scope.

dendibakh commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
````diff
@@ -0,0 +1,32 @@
+    ...
+}
+```
+The `ZoneScoped` macro will create an object on the stack with an associated timestamp. Once the function exits, the object will record a new timestamp and will store this timing data along with a few other details, including the function name. The implementation has very little overhead, although care must be taken when profiling sections of the code like tight loops.
````
Can you please define the term "zone". For folks not familiar with Tracy or game dev, it might be confusing. Maybe just change the first sentence in the paragraph to "The `ZoneScoped` macro creates an object that will record the runtime activity of the code within its scope, also called "zone". At the zone entry, the current timestamp is captured. [...]"

wolfpld replied:
Zone is just a term that was introduced by RAD Telemetry, and I decided to use it as well.

dendibakh commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -0,0 +1,32 @@
+Tracy has two operation modes: it can store all the timing data until the profiler is connected to the application, or it can simply discard the data and start recording as soon as a profiler is connected. The latter is usually the preferred option, as it means the tracing code can be compiled into the code and will cause little to no overhead to the running program unless the profiler is attache ...
+
+The profiler is a separate application that attaches to a running program to record and display the profiling data. The profiler can be run on a separate machine so that it doesn't interfere with the running application. Here is an example of the profiler output:
+![tracy profiler](../../img/tracy/profiler.png){#fig:Tracy_Profiler width=90%}
```
The current image is hard to view. It is very small when rendered in PDF. On a computer, readers can zoom, but not with a paperback :)

I think we have two options: 1) cut the panels that are not important, and 2) split the panels into multiple images.

Also, is it possible to change the dark mode to light mode? I'm afraid it won't be rendered nicely on paper.

The rendered book (PDF) can be seen here:
https://github.com/dendibakh/perf-book/actions/runs/4275046123

theWatchmen replied:
I will re-do the images so that they are (hopefully) clearer. Not sure if the profiler has a light mode, I will investigate.

wolfpld replied:
There is no official light theme, but it can be enabled with the following shader hack: https://github.com/wolfpld/tracy/issues/487

theWatchmen replied:
@wolfpld Thanks, I managed to redo all the screenshots except the source + assembly one. I tried to capture the TestCPU example from the Tracy repo - as admin and in debug mode - but I don't get the assembler. I have an AMD cpu. Do I need to install another library or do I need to enable some option when building Tracy locally? I am on Windows.

wolfpld replied:
The test application was not intended to show the disassembly. The manual outlines steps necessary to make this work.

dendibakh commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -0,0 +1,32 @@
+If debug symbols are available, Tracy can also display hot spots in the source code and related assembly:
+![tracy profiler](../../img/tracy/profiler3.png){#fig:Tracy_Profiler width=90%}
+
+Tracy provides many other features, too many to cover in this overview. They include:
```
Does Tracy have a "flight recorder" mode? I.e. visualize the data on the fly? Apple Instruments has that feature.

theWatchmen replied:
Yes, it's described in a previous paragraph: "The profiler is a separate application that attaches to a running program to record and display the live profiling data."

dendibakh reviewed (commented):
Hi Marco, it's a very good start.
Please find my comments below.

wolfpld commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -0,0 +1,40 @@
+## Tracing profilers
+Most of the tools explored so far fall under the category of sampling profilers. These are great when you want to identify hot-spots in your code, but in some cases they might not provide enough granularity for analysis. Depending on the profiler sampling frequency and the behavior of your program, most functions could be fast enough that they don't show up in a profiler. In some scenarios you mi ...
```
I'm not sure it's worth mentioning 30 FPS anymore. Monitors with 144 Hz refresh rates have been available for some time now, and with a 144 FPS target, we can show a slightly more dramatic 6.9 ms frame time requirement.

wolfpld commented:
Tracy is used for many applications other than making games. Examples include software for cloud streaming, medical microscopy, CAD, military simulation, and so on. Well-known companies outside the gaming industry that use Tracy include NVIDIA, Adobe, Amazon, Mozilla, TomTom, Netflix, Boston Dynamics, and CERN.

I do not agree with the statement that Tracy is specialized and VTune is for general use or a "go-to" solution. This seems to me to be heavily biased by what you are already familiar with. I see VTune as a highly specialized profiler for the following reasons:
- You would use it to do deep analysis of microarchitectural performance counters that are otherwise unavailable in other profilers.
- If you don't have an Intel CPU, you only get access to OS-level profiling features. Nothing provided by the CPU can be used.

In terms of sampling profiling, Tracy is both worse (because you can't attach to or launch an external process, or because the sampling is hard to find due to UI evolution) and better (for many reasons, for instance, see https://youtu.be/_hU7vw00MZ4?t=699 for a direct comparison to VTune) than the competition.

The main difference is that Tracy is also an instrumentation-based profiler. The main thing about working with this type of data, as opposed to sampling, is that you can clearly see when your otherwise fast function is occasionally slow. You can then drill down into what is happening in those rare cases to identify the cause.

dendibakh commented:
Ok, maybe I'm wrong and Tracy is a "go-to" tool for profiling. Don't want to debate as I haven't seen the data on what people usually use. 
This is a book about tuning SW for modern CPUs, so at the center is CPU performance, not system performance. We need insights into CPU microarchitecture.
But let me ask you @wolfpld, do you position Tracy as a competitor and a superior product to Vtune? This is an honest question, I know Vtune has flaws and in general, I'm a big proponent of open-source software. Don't get me wrong, I'm just genuinely trying to understand this. And I wish only the best for Tracy.
BTW, on AMD systems you would use UProf, not Vtune, on Apple platforms you would use Instruments. All those products are not perfect, but they have a specific purpose.

wolfpld commented:
>  This is a book about tuning SW for modern CPUs, so at the center is CPU performance, not system performance. We need insights into CPU microarchitecture.

You cannot look at the performance of your application without taking into account everything else that is going on in the system. See https://youtu.be/uJkrFgriuOo?t=72

> But let me ask you @wolfpld, do you position Tracy as a competitor and a superior product to Vtune? 

I had no need to use VTune for a long time, because to me Tracy *is* superior. I don't follow how they develop the product, so it's hard for me to say anything about it. It certainly has its uses, and if you want to take a deep dive into various specialized perf counters, VTune is the only way to go.

> BTW, on AMD systems you would use UProf

I could not use uProf at all. Any profiling session longer than a few seconds would lock the UI.

theWatchmen commented:
> @theWatchmen, should we rename this section as something like "Specialized profilers"?

I was thinking something like "Manual Instrumentation Profilers" or simply "Instrumentation Profilers" might be a better description.

dendibakh commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
````diff
@@ -0,0 +1,97 @@
+}
+```
+
+A [full example](https://github.com/wolfpld/tracy/tree/master/examples/ToyPathTracer) that shows how to use the Tracy API is provided in the official repository.
````
Please describe in 1-2 sentences what this test example is doing. It will be easier for readers to understand images.

theWatchmen replied:
Done

dendibakh commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
````diff
@@ -0,0 +1,97 @@
+```c++
+#include "tracy/Tracy.hpp"
+
+void Trace()
````
Is all the code needed? I can see you show the function names in the images later, but does it really adds value? I can't see you talk about `Trace`, `Scatter` and `DrawTest` functions.

Can we maybe write something like:

```cpp
#include "tracy/Tracy.hpp"
void RenderFrame() {
  ZoneScoped;
  for (...) {
    TraceRowJob();
  }
  FrameMark; // marks the end of the frame?
}

void TraceRowJob() {
  ZoneScoped;
  // ...
}
```
It will be slightly different from the original, but IMHO captures the same idea.

theWatchmen replied:
Done

dendibakh reviewed (commented):
Thanks Marco, light-mode images look much better (for the book)!

dendibakh commented:
> > @theWatchmen, should we rename this section as something like "Specialized profilers"?
> 
> I was thinking something like "Manual Instrumentation Profilers" or simply "Instrumentation Profilers" might be a better description.

Just want to make sure I understand it correctly: zone statistics (call counts, time, histogram) are exact because Tracy traces EVERY zone entry/exit, but system-level data and source-code-level data are sampled.
If we use the title "Instrumentation profilers", will it confuse people into thinking that Tracy instruments all the code (recursively) inside a zone? While it's just a marker API, that is recognized by Tracy.

theWatchmen commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -0,0 +1,97 @@
+
+The `ZoneScoped` macro creates an object on the stack that will record the runtime activity of the code within its scope. Tracy refers to this scope as "zone". At the zone entry, the current timestamp is captured. Once the function exits, the object will record a new timestamp and will store this timing data along with a few other details, including the function name. The implementation has very  ...
+
+Tracy has two operation modes: it can store all the timing data until the profiler is connected to the application, or it can simply discard the data and start recording only when a profiler is connected. The latter is usually the preferred option, as it means the tracing code can be compiled into the application and it will cause little to no overhead to the running program unless the profiler i ...
```
When the "on demand" option is active, all the profiling data is discarded until the profiler is connected. I have tried to clarify the paragraph :)

theWatchmen commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -0,0 +1,97 @@
+
+The profiler is a separate application that connects to a running application to record and display the live profiling data. The profiler can be run on a separate machine so that it doesn't interfere with the running application. Here is an example of the profiler output:
+
+![Tracy main view](../../img/tracy/profiler_main.png){#fig:Tracy_Main_View width=90%}
```
Done

theWatchmen commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -0,0 +1,97 @@
+
+Tracy has two operation modes: it can store all the timing data until the profiler is connected to the application, or it can simply discard the data and start recording only when a profiler is connected. The latter is usually the preferred option, as it means the tracing code can be compiled into the application and it will cause little to no overhead to the running program unless the profiler i ...
+
+The profiler is a separate application that connects to a running application to record and display the live profiling data. The profiler can be run on a separate machine so that it doesn't interfere with the running application. Here is an example of the profiler output:
```
That's right, the overhead for profiling the code is still present.

theWatchmen commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -0,0 +1,97 @@
+
+The profiler provides plenty of details. The main view shows the graph for all the zones that were active during a given frame. A histogram for displaying the times for all the recorded frames is displayed above the main view, which makes it easier to spot a long running frame (the red bar in the picture) that could cause stutter.
+
+![Tracy frame time view](../../img/tracy/profiler_frame_time.png){#fig:Tracy_Frame_Time_View width=90%}
```
I have kept the time histogram separate for clarity, but I have grouped the other four images. Let me know if it's readable :)

dendibakh commented:
Thanks Marco, I will review your updates in a couple of days.
BTW, I plan a new section for marker APIs (in progress), which showcases libpfm, which is not as fancy as Tracy, it's a wrapper around Linux perf_events. https://github.com/dendibakh/perf-book/pull/20
Since the concept of marker APIs is already covered, maybe we should position Tracy as a specialized/hybrid profiler and focus on its features for game developers for example?

theWatchmen commented:
> Since the concept of marker APIs is already covered, maybe we should position Tracy as a specialized/hybrid profiler and focus on its features for game developers for example?

Sure, that sounds good.

dendibakh commented:
@theWatchmen, I made a few edits, mostly cosmetic. I also left `[TODO]` comments in the text, please address.

dendibakh commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -25,43 +25,43 @@ void RenderFrame() {
 Right above the main panel, there is a histogram that displays the times for all the recorded frames, see figure @fig:Tracy_Frame_Time_View. It makes it easier to spot a long running frame that could cause stutter. It's not hard to spot a frame in the middle that takes twice as much time as an average frame.
 
-*[TODO:] @Marco, Can we also show the slow frame on the timeline that is right below this histogram?*
+*[TODO:] @Marco, Can we also show the slow frame on the timeline that is right below this histogram? @Denis the frame time view is taken from a different application. The path tracer example only runs for a few frames and would be hard to show this occurence*
```
Ok, got it. Then, I think, we should write about it explicitly not to trick readers :)
Or, if we have access to the code, maybe we could introduce this slow frame artificially? E.g. 
```
if (frameNumber == 5)
  sleep(100ms);
```
... instead of `sleep`, better to write some dummy `doExtensiveWork` func to make it seem real.

dendibakh replied:
> The path tracer example is not a "real time" application, so each frame already takes 100+ ms and it will be hard to spot a slow frame in the frame time view. Unfortunately I don't have another application with enough zones to show both a slow frame and have enough details for the other windows in Tracy.
> 
> I appreciate wanting to have a coherent narrative and show readers how to work through the slow frame case, but imho we are giving enough details about the tool for people to understand how they would debug such a case :)

Marco, is it possible to introduce an artificial random delay in the source code of the path tracer program?

theWatchmen replied:
Yes, we can do that.

dendibakh replied:
Can we please try introducing this random delay? 
So far I think this is one of the cool features (finding slow frames -> finding slow zones -> finding similar slow zones) in Tracy and it would be good to have a consistent story around it.

theWatchmen replied:
Ok, I will add a random call to a function `doExtraWork` that we'll then be highlighted in the text and the screenshots. Is that what you had in mind?

dendibakh replied:
Yes, please

dendibakh commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -25,43 +25,43 @@ void RenderFrame() {
+The Zone Info window (2) shows the details for a zone and how much of the execution time is due to the zone itself or its children. In this example, execution of the `TraceRowJob` function itself (without it's callees) in the current zone takes only 30.96 ms, which is 3.32% of the whole zone. The time for the zone including the child zones is 932.64 ms. This view also shows the source file and li ...
 
-Finally, in area 4 we can analyze the statistics for the data recorded, including the total time a given function function was active, how many times it was invoked, etc. It's also possible to select a time range in the main view an analyze the statistics only for that time interval. *[TODO:] @Marco, yellow lines on the histogram are barely visible, is it possible to use a different color?*
+Clicking on the Statistics button will display the Find Zone window (3). Here we can see the time histogram for a zone. This is particularly useful to determine how much variation there is when executing a function. This view also provides other data points, including the mean, median and standard deviation for the inspected zone.
```
I think we need to put emphasis here on the histogram, as this feature is not present in other tools (at far as I'm aware).
I guess it is very useful when you want to know if there are similar slow zone instances. You can then navigate to them and see what is common to all the slow zones, which should help you root cause performance issues.

dendibakh reviewed (commented):
Thank you @theWatchmen,
There are a few minor things to improve but only one high-level comment from my side: if possible, let's introduce this real slow frame (or two) in the path tracer to make a consistent story around it.

dendibakh commented:
@theWatchmen , are you waiting for my review? I thought you were still working on it. :)

theWatchmen commented:
Hey Denis, the last commit addresses your comments about artificially adding a slow zone and show that in the chapter. Sorry, the commit message doesn't make that clear. I don't have further changes for now :)

dendibakh commented on img/tracy/profiler_frame_time.png:
Can you please add an image with info about that slow frame once you point a mouse to that slow frame? E.g.:
![image](https://github.com/dendibakh/perf-book/assets/4634056/31a01ff7-f9bf-4ac4-a09d-829060086594)
Also, I wonder... in the description of ToyPathTracer it says it renders only 10 frames. Why the image shows more than 100? :) Is it from a different program?
Btw, we can extend ToyPathTracer to run for more frames if needed.

theWatchmen replied:
Added :) I modified the program to use a lower resolution so that we can clearly see the slow frames.

dendibakh reviewed (commented):
Marco, I added a few more comments.
Really want to make it look coherent, i.e. to have the code and images from the same program.

theWatchmen commented:
Hey Denis, apologies for the delay in getting back to you! I should have addressed all the latest comment and I have updated the images :) Let me know if you have any further notes.

theWatchmen commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -0,0 +1,83 @@
+
+The Zone Info window (2) shows the details for a zone and how much of the execution time is due to the zone itself or its children. In this example, execution of the `TraceRowJob` function itself (without it's callees) in the current zone takes only 654.99 us, which is 1.74% of the whole zone. The time for the zone including the child zones is 37.64 ms. This view also shows the source file and li ...
+
+Clicking on the "Statistics" button will display the Find Zone window (3). Here we can see the time histogram for a zone. This is particularly useful to determine how much variation there is when executing a function. From this view you can select one of the slow zones. This will update the Zone Info window (2) with the details of that zone instance and by clicking the "Zoom To zone" button, the  ...
```
I presented absolute values in the text, as tracy doesn't report percentile stats for the zone histogram (or at least, none that I could see). It's time consuming to compute this manually as I would need to manually inspect each bin :)

theWatchmen commented on img/tracy/profiler_main_crop_3.png:
Done :)

theWatchmen commented on chapters/8-Overview-Of-Performance-Analysis-Tools/8-7 Tracy.md:
```diff
@@ -0,0 +1,83 @@
+void TraceRowJob() {
+  ZoneScoped;
+  // ...
+}
```
Done :)

dendibakh commented:
Cool! Thanks Marco, it looks great.
Leave it to me. I will make some cosmetic changes and show you the final version for review.
One question: on "Figure 50: Tracy frame time view", which bar corresponds to frame 101? Maybe draw an arrow...
Also, the timeline on that image looks confusing to me... 
What do those numbers mean? :)
![image](https://github.com/dendibakh/perf-book/assets/4634056/0e92067c-8f2f-4458-af08-7baebe7ae897)

dendibakh commented:
I've made some changes. @theWatchmen, please review. The rendered PDF book is here:
https://github.com/dendibakh/perf-book/actions/runs/5557247288?pr=19
P.S. Sorry for the forced push. I rebased on top of the current main branch.

theWatchmen commented:
> One question: on "Figure 50: Tracy frame time view", which bar corresponds to frame 101? Maybe draw an arrow...
It's for the last bar in the graph. I have added a green rectangle to highlight it.

> Also, the timeline on that image looks confusing to me... What do those numbers mean? :) 
Those numbers are for the timeline below. I'll remove it.

> P.S. Marco, can you please provide details of the machine you used? CPU, OS, compiler.
Done.

theWatchmen commented:
I read the section on the PDF version, looks good to me :)

dendibakh commented:
> I read the section on the PDF version, looks good to me :)

Awesome! Thanks, merging it...
